### PR TITLE
first pass at iotm categorization

### PR DESCRIPTION
**Pull request by docrostov**
alright so i am starting out a big IOTM thing. 

i haven't really looked into how to get it hooked into everything you've gotten operational, sam, so it doesn't connect into greenbox.ts at all yet. but this is basically just staging the wealth of iotm data we need in a format that is (hopefully) convenient for generating what we need to get greenboxen populated. every iotm entry (tracked by year and month) will generate something like this:

```js
export interface IOTM {
    storeItem: Item;
    boundItem?: Item | Familiar;
    have: boolean;
    count: number;
}
```

the logic when transforming this into box color is that it'll be green if have is true, yellow if have is false but count is > 0, and white if count = 0 and have is false. my hope is that the way i set this up should in theory make it pretty easy to stage this data however the hell is most convenient for the existing scaffolding. but i am happy to take suggestions. 

also, though, please do not ask me to rewrite this all unless it can be done very easily without much typing. because dear fucking god this took so long. i cannot believe how many goddamn IOTMs exist in this game. what the everloving fuck, i swear to god. what on earth.

### CODE REVIEW

File: packages/greenbox-script/src/iotms.ts [hunk 1]
```diff
@@ -0,0 +1,474 @@
+import {availableAmount, Familiar, xpath, haveFamiliar, Item, Skill, visitUrl, floristAvailable} from "kolmafia";
+import {  $familiar, get, $item, $items, $skill, sum, getFoldGroup, haveInCampground, have as _have } from "libram";
+import { BooleanProperty } from "libram/dist/propertyTypes";
+
+// Exporting constants for months and years. 0 is for special casing.
+const months = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12] as const;
```
Comment (gausie) [inline packages/greenbox-script/src/iotms.ts:6]: Lousy smarch weather

File: packages/greenbox-script/src/iotms.ts [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        10: recordFamiliarIOTM($familiar`Vampire Vintner`),
+        11: recordItemIOTM($item`packaged Daylight Shavings Helmet`, $item`Daylight Shavings Helmet`),
+        12: recordWorkshedIOTM($item`packaged cold medicine cabinet`, $item`cold medicine cabinet`),
+    },
+    2022: {
+        1: recordPreferenceIOTM($item`undrilled cosmic bowling ball`,"hasCosmicBowlingBall"),
+        2: recordItemIOTM($item`combat lover's locket lockbox`, $item`combat lover's locket`),
+        3: recordFamiliarIOTM($familiar`Grey Goose`),
+        // I am not sure this will work for the umbrella honestly.
+        4: recordFoldableIOTM($item`undamaged Unbreakable Umbrella`),
```
Comment (gausie) [inline packages/greenbox-script/src/iotms.ts:468]: It's not foldable, it's just a single item stored as a single item id, so it's just an itemIoTM

**Comment by gausie** [comment]
This is good. We want to split this out slightly, moving the descriptive data into greenbox-data (so that greenbox-web can use it too!) and then defining state predicates that take a definition from greenbox-data and return whether the user has that defined item. I hope that makes sense - we want to have a definition of what each iotm is live seperately from the code that knows how to work out if you have each iotm (because the former is useful to both parts of the application, the latter is only useful (and indeed only makes sense) in the part of the application that runs in mafia

**Comment by docrostov** [comment]
cool! yep i think that makes sense. to play this back, what we need is:

**1: an iotms.ts file in greenbox-data/data that is formatted like so**

```js
export default [
   { id: "012004", name: "Dark Jill-O-Lantern", image: "jilldark" },
   ...
   { id: "082022", name: "unopened tiny stillsuit", image: "stillsuit_box" },
] as const;

**2: an iotms.ts file in greenbox-data/lib similar to classes.ts loading the above data**

**3: within greenbox.ts, something that accesses MrStoreMonthly and populates a big RawIOTM map**

have i unpacked the task properly or is there something major i'm missing here?

**Comment by gausie** [comment]
> cool! yep i think that makes sense. to play this back, what we need is:
> 
> **1: an iotms.ts file in greenbox-data/data that is formatted like so**
> 
> ```js
> export default [
>    { id: "012004", name: "Dark Jill-O-Lantern", image: "jilldark" },
>    ...
>    { id: "082022", name: "unopened tiny stillsuit", image: "stillsuit_box" },
> ] as const;
> ```
> 
> **2: an iotms.ts file in greenbox-data/lib similar to classes.ts loading the above data**
> 
> **3: within greenbox.ts, something that accesses MrStoreMonthly and populates a big RawIOTM map**
> 
> have i unpacked the task properly or is there something major i'm missing here?

No this is right! Making that file in greenbox-data then something in greenbox-script defining a callback for each line in there works.

File: build/config.gypi [hunk 3]
```diff
@@ -0,0 +1,364 @@
+# Do not edit. File was generated by node-gyp's "configure" step
```
Comment (gausie) [inline build/config.gypi:1]: Pls gitignore

File: packages/greenbox-script/src/generateIotms.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
+import { Item, print, toInt } from "kolmafia";
```
Comment (gausie) [inline packages/greenbox-script/src/generateIotms.ts:1]: This probably wants to live elsewhere - we could make a greenbox-tools package for this stuff maybe

Comment (gausie) [reply packages/greenbox-script/src/generateIotms.ts:1]: With its own build.mjs

Comment (docrostov) [reply packages/greenbox-script/src/generateIotms.ts:1]: oops… yeah sorry i meant to delete this before the push. my b

File: build/config.gypi (hunk 3 above)
Comment (docrostov) [reply build/config.gypi:1]: neverrrr

Comment (docrostov) [reply build/config.gypi:1]: (ok)

**Comment by gausie** [comment]
Ah I shouldn't have merged this since we deploy from main. I'm going to refactor this a bunch as no insult to yourself beloved Scotch

**Pull request merged**